provider/oci/common/client.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package common
+
+import (
+	"context"
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/instance"
+	"github.com/juju/juju/network"
+
+	ociCommon "github.com/oracle/oci-go-sdk/common"
+	ociCore "github.com/oracle/oci-go-sdk/core"
+	ociIdentity "github.com/oracle/oci-go-sdk/identity"
+)
+
+type jujuConfigProvider struct {
+	key            []byte
+	keyFingerprint string
+	passphrase     string
+	tenancyOCID    string
+	userOCID       string
+	region         string
+}
+
+type ociClient struct {
+	ociCore.ComputeClient
+	ociCore.BlockstorageClient
+	ociCore.VirtualNetworkClient
+	ociIdentity.IdentityClient
+
+	ociCommon.ConfigurationProvider
+}
+
+// NewJujuConfigProvider returns a new ociCommon.ConfigurationProvider instance
+func NewJujuConfigProvider(user, tenant string, key []byte, fingerprint, passphrase, region string) ociCommon.ConfigurationProvider {
+	return &jujuConfigProvider{
+		key:            key,
+		keyFingerprint: fingerprint,
+		passphrase:     passphrase,
+		tenancyOCID:    tenant,
+		userOCID:       user,
+		region:         region,
+	}
+}
+
+func NewOciClient(provider ociCommon.ConfigurationProvider) (ApiClient, error) {
+	computeClient, err := ociCore.NewComputeClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	blockStorage, err := ociCore.NewBlockstorageClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	virtualNetwork, err := ociCore.NewVirtualNetworkClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	ident, err := ociIdentity.NewIdentityClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &ociClient{
+		computeClient,
+		blockStorage,
+		virtualNetwork,
+		ident,
+		provider,
+	}, nil
+}
+
+// Ping validates that the client can access the OCI API successfully
+func (o *ociClient) Ping() error {
```

Comment:
Given you implemented ociClient.Ping() is there a reason it isn't being used in EnvironProvider.Ping() ?

Reply:
The signature of the method in EnvironProvider implied  that the ping would be done against an endpoint. In OCI, the endpoint is static, hardcoded in the SDK, and it is based on region. So after setting the region in the config, the ```endpoint``` arg passed into the function would be completely ignored, and the ping would go to whatever endpoint is configured in the sdk for the region we selected. It felt like a big fat lie to implement it. A ```Ping()``` is done in ```PrepareForBootstrap()```. If you think its fine to ignore the ```endpoint``` argument of ```EnvironProvider.Ping()```, I can add it.

Reply:
We do a Ping() once we've configured something to make sure that you didn't get it wrong, and we really can contact the provider there (be it, MAAS, Openstack, special region in AWS, etc).
It does feel like EnvironProvider.Ping() could check that the given URL is correct, and then call the underlying Ping(). But its just a "are the credentials considered valid at the endpoint that we're trying to connect to". It doesn't have to be ociClient.Ping() as long as it validates things are working.